samcli/hook_packages/terraform/hooks/prepare.py

```diff
@@ -93,6 +108,29 @@ def prepare(params: dict) -> dict:
         raise PrepareHookException(f"Unable to create directory {output_dir_path}") from e
 
 
+def _update_resources_paths(cfn_resources: Dict[str, Any], terraform_application_dir: str) -> None:
+    """
+    As Sam Cli and terraform handles the relative paths differently. Sam Cli handles the relative paths to be relative
+    to the template, but terraform handles them to be relative to the project root directory. This Function purpose is
+    to update the CFN resources paths to be absolute paths, and change relative paths to be relative to the terraform
+    application root directory.
+
+    Parameters
+    ----------
+    cfn_resources: dict
+        CloudFormation resources
+    terraform_application_dir: str
+        The terraform application root directory where all paths will be relative to it
+    """
+    resources_attributes_to_be_updated = {CFN_AWS_LAMBDA_FUNCTION: ["Code"]}
```

Comment:
yes we do it while generating the CFN metadata, but I though it is better to have a separate method to update the resource paths just before we writing it, as we are going to add more logic after the CFN generation, like the updating it using the sam metadata resources, and also the resources linking, and I do not want to add the paths updating every where in our code.

Reply:
Ideally, I would like us to be building on top of these things instead of recreating them. Having a single location is super beneficial as we grow. I will leave it up to you, but don't think its a good idea to be duplicating this throughout the code base. Using what we already have (in some form or refactoring parts) is going to make these things much more maintainable.

Reply:
to make sure that I do not miss anything, are you talking about this part in the code:

https://github.com/aws/aws-sam-cli/blob/a4940ac555054f1d27d9ffa11548d2fd6a8b0a0c/samcli/hook_packages/terraform/hooks/prepare.py#L270-L280

If yes, so I do not think it is duplication, in this part we do the mapping, and then in this method we do the path updates (it can be for subset of these parameters, or for other parameters as well). The idea is we are going to add more logic to the prepare hook for other resources mapping, and the SAM Metadata mapping, so to avoid doing the path updates every where in the code, I though it is better to be done in one place

Reply:
as per our discussion, I miss-understood your comment at the beginning. Your comment is mainly to reuse some of the existing utilities to update the Template resources paths, and you referred to something like this https://github.com/aws/aws-sam-cli/blob/a4940ac555054f1d27d9ffa11548d2fd6a8b0a0c/samcli/lib/build/app_builder.py#L351-L364

This method could not be used in our case, as here we will update different resources attributes like as we have now the Lambda resources Code Uri, but also we will use it to update API Gateway resources definition body paths as an example. Also, we can update metadata properties, like the DockerFile path in case of Image Lambda functions. I also do not want to update this method to do what we need to avoid introduce issues to `sam build` command.

I tried to find some other utility that can be used, but I did not find.